compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1760,7 +1760,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let coerce_to = expected
                 .to_option(self)
                 .and_then(|uty| match *self.try_structurally_resolve_type(expr.span, uty).kind() {
-                    ty::Array(ty, _) | ty::Slice(ty) => Some(ty),
+                    // Avoid using a type variable as the coerce_to type, as it may resolve
+                    // during the first coercion instead of being based on the common type.
+                    ty::Array(ty, _) | ty::Slice(ty) if !ty.is_ty_var() => Some(ty),
```

Comment:
u need to structurally resolve the `ty` again before this check :thinking: `try_structurally_resolve_type` only structurally resolves the outermost type constructor